dqvm/src/addresses.lisp

```diff
@@ -0,0 +1,282 @@
+         (permutation (permutation global-addresses)))
+
+    (assert (not (minusp offset)))
+
+    (when (< offset number-of-addresses)
+      (let ((base-address (if (< offset (* blocks-per-process block-size))
+                              (* rank blocks-per-process block-size)
+                              (* block-size
+                                 (+ rank
+                                    (* blocks-per-process
+                                       (1- number-of-processes)))))))
+        (apply-qubit-permutation permutation (+ offset base-address))))))
+
+(defun %get-address-ranges (addresses)
+  (let* ((rank (rank addresses))
+         (number-of-processes (number-of-processes addresses))
+         (block-size (block-size addresses))
+         (blocks-per-process (blocks-per-process addresses))
+         (remainder-blocks (remainder-blocks addresses))
+         (k (* block-size blocks-per-process))
+         (from-0 (* k rank))
+         (end-0 (* k (1+ rank))))
+
+    (if (< rank remainder-blocks)
+        (let* ((n-times-k (* k number-of-processes))
+               (from-1 (+ n-times-k (* block-size rank)))
+               (end-1 (+ n-times-k (* block-size (1+ rank)))))
+          (values from-0 end-0 from-1 end-1))
+
+        (values from-0 end-0))))
+
+(defmacro do-addresses ((var addresses &optional result) &body body)
+  "Iterate over ADDRESSES using VAR.
+
+The addresses are generated on the fly based on the rank of ADDRESSES and the permutation."
+  (alexandria:once-only (addresses)
+    (alexandria:with-gensyms (start-0 stop-0 start-1 stop-1 %body permutation index)
+      `(multiple-value-bind (,start-0 ,stop-0 ,start-1 ,stop-1)
+           (%get-address-ranges ,addresses)
+
+         (flet ((,%body (,var)          ; XXX (declaim (inline ,%body)) ?
+                  ,@body))
+
+           (let ((,permutation (permutation ,addresses)))
+
+             (loop :for ,index :from ,start-0 :below ,stop-0
+                   :for ,var := (apply-qubit-permutation ,permutation ,index)
+                   :do (,%body ,var))
+
+             (when (and ,start-1 ,stop-1)
+               (loop :for ,index :from ,start-1 :below ,stop-1
+                     :for ,var := (apply-qubit-permutation ,permutation ,index)
+                     :do (,%body ,var)))
+
+             ,result))))))
+
+(defmethod print-object ((addresses addresses) stream)
+  (let ((*print-readably* nil)
+        (*print-pretty* nil))
+
+    (print-unreadable-object (addresses stream :type t :identity t)
+
+      (format stream "~{~A ~A~^ ~}"
+              (list (prin1-to-string :rank) (rank addresses)
+                    (prin1-to-string :global-addresses) (global-addresses addresses)))
+
+      (when *print-addresses*
+        (let (address-list)
+          (do-addresses (address addresses)
+            (push address address-list))
+          (when address-list
+            (format stream " ~A (~{~D~^ ~})"
+                    (prin1-to-string :addresses)
```

Comment:
Right. I was using `prin1` to (1) keep the colon in the printed symbol and (2) preserve `*print-case*`. I've since dropped `prin1` and changed the format string to use ~S instead of ~A, which satisfies the two desired points mentioned above.